### PR TITLE
Unit task

### PR DESCRIPTION
**Pull request by nastydrew**

### CODE REVIEW

File: Unit/Unit.cpp [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+const std::string& Unit::getName() const {
+	return name;
+}
+
+void Unit::addHitPoints(int hp) {
+	ensureIsAlive();
+
+	if ( hitPointsLimit - hitPoints < hp ) {
+		hitPoints = hitPointsLimit;
+	} else {
```
Comment (igordotsenko) [inline Unit/Unit.cpp]:
```
if ( ... ) {
...
return 
}
...
```

File: Unit/Unit.cpp [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+	}
+}
+
+void Unit::takeDamage(int dmg) {
+	ensureIsAlive();
+
+	hitPoints -= dmg;
+	if ( hitPoints <= 0 ) {
+		hitPoints = 0;
+		throw UnitIsDeadException();
```
Comment (igordotsenko) [inline Unit/Unit.cpp]: I don't think that you have to throw it here.

File: Unit/Unit.cpp [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+	}
+}
+
+void Unit::attack(Unit& enemy) {
+	enemy.takeDamage(damage);
+	cout << getName() << " taked " << damage << " damage." << endl;
+
+	ensureIsAlive();
```
Comment (igordotsenko) [inline Unit/Unit.cpp]: Why here?
You should do it before your enemy takes damage

Comment (nastydrew) [inline Unit/Unit.cpp]: I'm agree. But  "takeDamage" already has this check. Why should replace it before `enemy.takeDamage(damage);` ?

File: Unit/Unit.cpp (hunk 1 above)
Comment (nastydrew) [inline Unit/Unit.cpp]: What are doing "return" in void function?

Comment (igordotsenko) [inline Unit/Unit.cpp]: Come on, 1st week's question...
Breaks the method executing

File: Unit/Unit.cpp (hunk 3 above)
Comment (igordotsenko) [inline Unit/Unit.cpp]: But what if attacking unit is already dead? Zombie attack? )

File: Unit/Unit.cpp [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+#include "Unit.h"
+
+using namespace std;
+
+void Unit::ensureIsAlive() {
+	if ( hitPoints == 0 ) {
+		throw UnitIsDeadException();
+	}
+}
+
```
Comment (igordotsenko) [inline Unit/Unit.cpp:11]: Code formatting. Why 8 spaces?

File: Unit/Unit.cpp [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+void Unit::takeDamage(int dmg) {
+	hitPoints -= dmg;
+	if ( hitPoints <= 0 ) {
+		hitPoints = 0;
+	}
+}
+
+void Unit::attack(Unit& enemy) {
+	ensureIsAlive();
+	enemy.ensureIsAlive();
```
Comment (igordotsenko) [inline Unit/Unit.cpp]: It's better to put this in "take damage". Handle exception thrown by takeDamage in current method

File: Unit/Unit.cpp [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+	enemy.takeDamage(damage);
+	cout << getName() << " taked " << damage << " damage. " << endl;
+
+	enemy.counterAttack(*this);
+}
+
+void Unit::counterAttack(Unit& enemy) {
+	ensureIsAlive();
+	enemy.ensureIsAlive();
```
Comment (igordotsenko) [inline Unit/Unit.cpp]: If you put this in "take damage"

**Pull request merged**